code/game/machinery/computer/gulag_teleporter.dm

```diff
@@ -0,0 +1,159 @@
+
+	if(teleporter)
+		data["teleporter"] = teleporter
+		data["teleporter_location"] = "([teleporter.x], [teleporter.y], [teleporter.z])"
+		data["teleporter_lock"] = teleporter.locked
+		data["teleporter_state_open"] = teleporter.state_open
+	if(beacon)
+		data["beacon"] = beacon
+		data["beacon_location"] = "([beacon.x], [beacon.y], [beacon.z])"
+	if(id)
+		data["id"] = id
+		data["id_name"] = id.registered_name
+		data["goal"] = id.goal
+	data["can_teleport"] = can_teleport
+
+	return data
+
+/obj/machinery/computer/gulag_teleporter_computer/ui_act(action, list/params)
+	if(..())
+		return
+	if(!allowed(usr))
+		usr << "<span class='warning'>Access denied.</span>"
+		return
+	switch(action)
+		if("scan_teleporter")
+			teleporter = findteleporter()
+		if("scan_beacon")
+			beacon = findbeacon()
+		if("handle_id")
+			if(id)
+				if(!usr.get_active_hand())
+					usr.put_in_hands(id)
+					id = null
+				else
+					id.loc = get_turf(src)
+					id = null
+			else
+				var/obj/item/I = usr.get_active_hand()
+				if(istype(I, /obj/item/weapon/card/id/prisoner))
+					if(!usr.drop_item())
+						return
+					I.loc = src
+					id = I
+		if("set_goal")
+			var/new_goal = input("Set the amount of points:", "Points", id.goal) as num|null
+			if(!isnum(new_goal))
+				return
+			if(!new_goal)
+				new_goal = default_goal
+			id.goal = Clamp(new_goal, 0, 1000) //maximum 1000 points
+		if("toggle_open")
+			if(teleporter.locked)
+				usr << "The teleporter is locked"
+				return
+			teleporter.toggle_open()
+		if("teleporter_lock")
+			if(teleporter.state_open)
+				usr << "Close the teleporter before locking!"
+				return
+			teleporter.locked = !teleporter.locked
+		if("teleport")
+			if(!teleporter || !beacon)
+				return
+			addtimer(src, "teleport", 5, FALSE, usr)
+
+/obj/machinery/computer/gulag_teleporter_computer/proc/scan_machinery()
+	teleporter = findteleporter()
+	beacon = findbeacon()
+
+/obj/machinery/computer/gulag_teleporter_computer/proc/findteleporter()
+	var/obj/machinery/gulag_teleporter/teleporterf = null
+
+	for(dir in cardinal)
+		teleporterf = locate(/obj/machinery/gulag_teleporter, get_step(src, dir))
+		if(teleporterf && teleporterf.is_operational())
+			return teleporterf
+
+/obj/machinery/computer/gulag_teleporter_computer/proc/findbeacon()
+	return locate(/obj/structure/gulag_beacon)
+
+/obj/machinery/computer/gulag_teleporter_computer/proc/teleport(mob/user)
+	log_game("[user]([user.ckey] teleported [prisoner]([prisoner.ckey]) to the Labor Camp ([beacon.x], [beacon.y], [beacon.z]) for [id.goal] points.")
+	teleporter.handle_prisoner(id, temporary_record)
+	playsound(loc, "sound/weapons/emitter.ogg", 50, 1)
+	prisoner.loc = get_turf(beacon)
+	prisoner.Weaken(2) // small travel dizziness
+	prisoner << "<span class='warning'>The teleportation makes you a little dizzy.</span>"
+	PoolOrNew(/obj/effect/particle_effect/sparks, prisoner.loc)
+	playsound(src.loc, "sparks", 50, 1)
+	if(teleporter.locked)
```

Comment:
this isn't really necessary, you don't gain anything by only setting it FALSE if it's TRUE, it's probably actually SLOWER to check